arch: Add support to imx7d m4 core and colibri_imx7d_m4 board

diegosueiro opened the pull request:
board/colibri_imx7d_m4: added board support to toradex colibri imx7d single board computer

This change basically adds support of a popular IMX7D processors from NXP, which has a ARM Cortex A7 Core and a Cortex M4 Core, this change is targeted to the second one and now is possible to select it as a supported SoC, Toradex has some attractive single board computers with industrial grade using this processor and this change also the basic support of the board and the NXP supplied HAL from FreeRTOS.

This support is intended to increase the options of heterogeneous processors in Zephyr SoC portfolio, adds the SoC and Board template is the first step only, so we divided the change in two steps, first provide a functional SoC support to perform build and image generation to be flashed into differtent RAMs (imx7d m4 core does not has built in flash) and the serial drivers and the next step is the IPM driver to communicate m4 core with a7 core, this approach was chosen to keep the changes not too large and hard to verify.

This changed was validated as follows:

- build the hello world sample under build folder with: $cmake -DBOARD=colibri_imx7d_m4 .. and make.

- the generated image can be loaded through a Jlink probe or with A7 running Linux + uBoot using the fatload command

More details can e found in the boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst file.

This is a rework of the PR [#5330](https://github.com/zephyrproject-rtos/zephyr/pull/5330)
Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

carlescufi commented:
@diegosueiro https://github.com/zephyrproject-rtos/zephyr/issues/3065 and https://github.com/zephyrproject-rtos/zephyr/pull/5960 FYI

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+
+The i.MX7 SoC is a Hybrid multi-core processor composed by Single/Dual Cortex A7
+core and Single Cortex M4 core.
+The Zephry was ported to run on the M4 core and the final goal is to have
```
Change to:

    Zephyr was ported to run on the M4 core.  In a later release, it will also communicate
    with the A7 core (running Linux) via RPmsg.

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+===================
+
+The Colibri iMX7D Computer on Module with Colibri Evaluation Board
+was tested with following pinmux controller configuration.
```
with the following

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+System Clock
+============
+
+The M4 Core is configured to run at 200 MHz clock.
```
at a 200 MHz clock speed.

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+Serial Port
+===========
+
+The iMX7D SoC has seven UARTs. Two is configured for the console and the
```
Two are configured

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+
+The Colibri iMX7D doesn't have QSPI flash for the M4 and it needs to be started by
+the A7 core. The A7 core is responsible to load the M4 binary application into the
+RAM, put the M4 in reset, set the M4 Program Counter and Stack Pointer and get the
```
add a comma:

    and Stack Pointer, and get the

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+done in the file ``boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.dts`` with "zephyr,flash" (when CONFIG_XIP=y) and "zephyr,sram"
+properties. The available configurations are:
+
+"zephyr,flash"
```
Since these are lines in the dts file, use a code block:

    .. code-block:: none

      "zephyr, flash"
      - &ddr_code

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+configured in the Zephyr compilation:
+
+.. code-block:: console
+  setenv bootm4 'ext2load mmc 0:1 $m4addr $m4fw && dcache flush && bootaux $m4addr'
```
need a blank line after the ``.. code-block`` directive and then all the lines within the code block need to be indented to the "c" in "code-block"

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+
+get-pc-sp.sh:
+.. code-block:: console
+  #!/bin/sh
```
need a blank line after the ``.. code-block`` directive, and all the code block lines need to be indented to the "c" in "code-block"

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+
+
+Get the SP and PC from firmware binary: ``./get-pc-sp.sh zephyr.bin``
+.. code-block:: console
```
same here, blank line and indentation

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+Plug in the J-Link into the board and PC and run the J-Link command line tool:
+
+.. code-block:: console
+  /usr/bin/JLinkExe -device Cortex-M4 -if JTAG -speed 4000 -autoconnect 1 -jtagconf -1,-1 -jlinkscriptfile iMX7D_Connect_CortexM4.JLinkScript
```
and here: blank line and indentation

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+  /usr/bin/JLinkExe -device Cortex-M4 -if JTAG -speed 4000 -autoconnect 1 -jtagconf -1,-1 -jlinkscriptfile iMX7D_Connect_CortexM4.JLinkScript
+
+The following steps are necessary to run the zephyr.bin:
+1 Put the M4 core in reset
```
for a numbered list, add a blank line before the list, and use a period after the number:

    The following steps are necessary to run the zephyr.bin:

    1. Put the M4 core in reset
    2. Load the binary ...

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+Issue the following commands inside J-Link commander:
+
+.. code-block:: console
+  w4 0x3039000C 0xAC
```
blank line and indenting

dbkinder commented on boards/arm/colibri_imx7d_m4/doc/colibri_imx7d_m4.rst:
```diff
@@ -0,0 +1,339 @@
+- `J-Link iMX7D Instructions`_
+
+
+.. zephyr-app-commands::
```
This seems out of place?  Does it below above in the debugging section?

diegosueiro commented:
@dbkinder,

Thanks for the documentation review. I already addressed your comments.

MaureenHelm commented on drivers/serial/uart_imx.c:
```diff
@@ -0,0 +1,575 @@
+
+static struct uart_imx_dev_data_t uart_imx_dev_data_1 = {
+	.baud_rate = CONFIG_UART_IMX_UART_1_BAUD_RATE,
+	.modem_mode = CONFIG_UART_IMX_UART_1_MODEM_MODE,
```
These values never get updated at runtime, so you can move them to the const config structure.

MaureenHelm commented on drivers/serial/uart_imx.c:
```diff
@@ -0,0 +1,575 @@
+
+static const struct uart_device_config uart_imx_dev_cfg_1 = {
+	.base = (uint8_t *)CONFIG_UART_IMX_UART_1_BASE_ADDRESS,
+	.sys_clk_freq = MHZ(100),
```
Why is each instance set to MHZ(100), MHZ(200), ...? Are you using sys_clk_freq anywhere?

MaureenHelm commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,223 @@
+			status = "disabled";
+		};
+
+		uart2: uart@30890000 {
```
The address should be 30870000

diegosueiro replied:
I think that the right address value is 0x30890000. Looking at the i.MX7D reference manual on page 198 it mentions 0x30870000 but on page 4534 it mentions 0x30890000. I tested with 0x30870000 and couldn't get the UART working.

MaureenHelm replied:
The cmsis header also uses 0x30890000, so it looks like you're right and there's an error in the reference manual. I'll let our docs team know.

MaureenHelm commented on ext/hal/nxp/imx/Kconfig:
```diff
@@ -0,0 +1,38 @@
+
+endif # UART_IMX 
+
+config HAS_IMX_DRC
```
s/DRC/RDC/

MaureenHelm commented on ext/hal/nxp/imx/devices/MCIMX7D/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+
+zephyr_sources(
+ system_MCIMX7D_M4.c
+ startup_MCIMX7D_M4.S
```
I don't think we need to build either of these. The zephyr cortex-m port already does all or most everything in startup_MCIMX7D_M4.S. The zephyr mpu driver does a lot of what's in system_MCIMX7D_M4.c, so you might just need to grab the cache init stuff and copy it to soc.c. You can do a follow up patch later to enable the existing mpu driver.

MaureenHelm commented on ext/hal/nxp/imx/drivers/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+
+
+
```
Remove the extra blank lines

MaureenHelm reviewed (changes requested):
Thanks @diegosueiro for picking this back up!

Please split up the changes into separate commits:
1. Import the nxp freertos bsp
2. Add the imx uart driver shim
3. Add the imx7d m4 soc
4. Add the colibri_imx7d_m4 board

cc: @MarekNovakNXP @PetrLukas @stanislav-poboril

MaureenHelm commented on ext/hal/nxp/imx/Kconfig:
```diff
@@ -0,0 +1,38 @@
+# Kconfig - IMX M4 Core SDK
+#
+# Copyright (c) 2016, Freescale Semiconductor, Inc.
```
Update the year and use your own copyright. If you want to assign copyright to Freescale/NXP, use the NXP copyright: `Copyright (c) 2018, NXP`

MaureenHelm commented on drivers/serial/Kconfig.imx:
```diff
@@ -0,0 +1,128 @@
+# Kconfig - iMX7 M4 UART
+#
+# Copyright (c) 2017, NXP
```
Same comment as before about the copyright

MaureenHelm commented on drivers/serial/Kconfig.imx:
```diff
@@ -0,0 +1,128 @@
+	select SERIAL_HAS_DRIVER
+	depends on HAS_IMX_HAL
+	help
+	This option enables the UART driver for NXP i.MX7
```
Indent with two spaces

MaureenHelm commented on drivers/serial/Kconfig.imx:
```diff
@@ -0,0 +1,128 @@
+      Enable support for UART1 port in the driver. Say y here
+      if you want to use UART1 device.
+
+config UART_IMX_UART_1_NAME
```
You can remove the _NAME configs and get the name from DTS

MaureenHelm commented on drivers/serial/uart_imx.c:
```diff
@@ -0,0 +1,544 @@
+	((UART_Type *)(DEV_CFG(dev))->base)
+
+/* Device data structure */
+struct uart_imx_dev_data_t {
```
This is really a config structure, not a data structure. Can you please rename it to `struct uart_imx_dev_config`

MaureenHelm commented on drivers/serial/uart_imx.c:
```diff
@@ -0,0 +1,544 @@
+static const struct uart_driver_api uart_imx_driver_api;
+
+static int uart_imx_poll_in(struct device *dev, unsigned char *c);
+static unsigned char uart_imx_poll_out(struct device *dev, unsigned char c);
```
you don't need these declarations

MaureenHelm commented on drivers/serial/uart_imx.c:
```diff
@@ -0,0 +1,544 @@
+	uart_init_config_t initConfig = {
+		.baudRate	= DEV_DATA(dev)->baud_rate,
+		.wordLength = uartWordLength8Bits,
+		.stopBitNum = uartStopBitNumOne,
```
Please fix the alignment

MaureenHelm commented on boards/arm/colibri_imx7d_m4/board.c:
```diff
@@ -0,0 +1,56 @@
+SYS_INIT(colibri_imx7d_init, PRE_KERNEL_1, 0);
+/*******************************************************************************
+ * EOF
+ ******************************************************************************/
```
Remove this

MaureenHelm commented on boards/arm/colibri_imx7d_m4/board.c:
```diff
@@ -0,0 +1,56 @@
+	dbg_uart_init();
+#endif /* CONFIG_UART_IMX_CONSOLE_DEBUG */
+
+		return 0;
```
Fix alignment

MaureenHelm commented on boards/arm/colibri_imx7d_m4/board.h:
```diff
@@ -0,0 +1,102 @@
+#define BOARD_I2C_IRQ_NUM                     I2C4_IRQn
+#define BOARD_I2C_HANDLER                     I2C4_Handler
+#define BOARD_I2C_FXAS21002_ADDR              (0x20)
+#define BOARD_I2C_FXOS8700_ADDR               (0x1E)
```
What are all the macros in this file for? Most look like things we can get from dts

MaureenHelm commented on boards/arm/colibri_imx7d_m4/gpio_pins.c:
```diff
@@ -0,0 +1,78 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
```
This file and its counterparts need to go in ext/ if we're going to use them

MaureenHelm reviewed (changes requested):
Some of the commit messages are missing a body, which will cause CI to fail.

It looks like you've merged unrelated commits. Please rebase your branch and force push to github, rather than merge.

diegosueiro commented:
@MaureenHelm,

Any further feedback on this review?

galak reviewed (changes requested):
In the commit that you import the SDK for imx - you need to update the commit message, take a look at the bottom of https://github.com/zephyrproject-rtos/zephyr/blob/master/CONTRIBUTING.rst to see examples of the format of the commit message.

galak reviewed (changes requested):
In the import of the SDK can you maintain the dir struct of utilities/inc & utilies/src

galak reviewed (changes requested):
Can you split out the CMakeLists.txt & KConfig additions to the hal/sdk import into their own patch.  The changes that add support for building a driver should be with that driver patch.  For example in ext/hal/nxp/imx/drivers/CMakeLists.txt:

+ zephyr_sources_ifdef(CONFIG_UART_IMX                uart_imx.c)
  
Should be part of the UART driver patch.

galak commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,223 @@
+			reg = <0x30860000 0x10000>;
+			interrupts = <26 0>;
+			interrupt-names = "status";
+			zephyr,irq-prio = <3>;
```
zephyr,irq-prio shouldnt be used, the priority value is the second cell in the interrupt prop.  So remove this and have:

interrupts = <26 3>;

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/soc.c:
```diff
@@ -0,0 +1,93 @@
+
+#ifdef CONFIG_UART_IMX_UART_2
+	/* We need to grasp board uart exclusively */
+	RDC_SetPdapAccess(RDC, rdcPdapUart2, 3 << (BOARD_DOMAIN_ID * 2), false,
```
need to cleanup magic number 3, probably a comment or something about why (BOARD_DOMAIN_ID * 2)

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/soc.c:
```diff
@@ -0,0 +1,93 @@
+static void nxp_mcimx7_uart_config(void)
+{
+
+#ifdef CONFIG_UART_IMX_UART_2
```
What about all the other UARTs?

diegosueiro replied:
At the moment I just tested the UART_2. As soon as I test the other UARTs I'll add them.

galak commented on ext/hal/nxp/imx/drivers/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+zephyr_include_directories(.)
+
+zephyr_sources(clock_freq.c)
+zephyr_sources_ifdef(CONFIG_GPIO_IMX                gpio_imx.c)
```
CONFIG_GPIO_IMX doesnt seem to exist anywhere.

galak commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,223 @@
+			compatible = "nxp,imx-uart";
+			reg = <0x30860000 0x10000>;
+			interrupts = <26 0>;
+			interrupt-names = "status";
```
remove names, since there is only one interrupt line, its not terribly useful.

galak commented on boards/arm/colibri_imx7d_m4/Kconfig.new:
```diff
@@ -0,0 +1,18 @@
+# Kconfig - Colibri iMX7D M4 board
```
should Kconfig.new exist?

galak reviewed (changes requested):
Various code cleanups.

galak reviewed (commented):
Do we need debug console support at all?  Since we have a proper uart driver and printf/scanf support in Zephyr?

Is there something else that DbgConsole is providing?

galak commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,223 @@
+			label = "DDR CODE";
+
+			clock-controller;
+			#clocks-cells = <3>;
```
why does ddr_code have a clock-controller; and #clock-cells.  They should probalby be removed (same for all the various memory nodes).

galak commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,223 @@
+			#clocks-cells = <3>;
+		};
+
+		ocram_pxp_code: code@00940000 {
```
Does it really make sense to have pxp & edpc used for memory/code storage for Zephyr?  It seems like TCM would be the default, and exposing OCRAM_S, OCRAM, and DDR would be more than enough.

diegosueiro replied:
As soon as pxp and edpc can be used for this purpose I thought that should be a good idea to have them exposed as well.

galak replied:
I'm not opposed, but assume that PXP & EDPC are meant to normally be used by the various graphics devices/drivers that they are associated with.  Also, I'd assume the performance of going across various SoC busses to those memory is going to be higher compared to TCM.  It just feels rare that PXP or EDPC memories would be what one would choose, but someone could tell me otherwise.

galak reviewed (changes requested):
Various comments related to memory regions.

What memory uses cases really make sense to support?  Seems like the default should be running out of TCM.

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/soc.h:
```diff
@@ -0,0 +1,36 @@
+#include "ccm_imx7d.h"
+#include "clock_freq.h"
+
+#define NO_DOMAIN_PERM 0
```
How about the following:

```
#define RDC_DOMAIN_PERM_NONE  (0x0)
#define RDC_DOMAIN_PERM_W     (0x1)
#define RDC_DOMAIN_PERM_R     (0x2)
#define RDC_DOMAIN_PERM_RW    (RDC_DOMAIN_PERM_W|RDC_DOMAIN_PERM_R)

#define RDC_DOMAIN_PERM(domain, perm) (perm << (domain * 2))
```

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/dts.fixup:
```diff
@@ -0,0 +1,48 @@
+#define CONFIG_UART_IMX_UART_1_NAME         NXP_IMX_UART_30860000_LABEL
```
Seems like you need a fixup for prio bits:

```
#define CONFIG_NUM_IRQ_PRIO_BITS                ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
```

galak commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,147 @@
+};
+
+&nvic {
+	arm,num-irqs = <127>;
```
remove arm,num-irqs  - its not used, or defined anywhere.

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/Kconfig.series:
```diff
@@ -0,0 +1,16 @@
+	select CPU_CORTEX_M4
+	select SOC_FAMILY_IMX
+	select CPU_HAS_SYSTICK
+	select CLOCK_CONTROL
```
select CPU_HAS_FPU

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,24 @@
+config SOC_SERIES
+	default mcimx7_m4
+
+config NUM_IRQ_PRIO_BITS
```
remove NUM_IRQ_PRIO_BITS, this should come from DT

galak commented on arch/arm/soc/nxp_imx/mcimx7_m4/soc.c:
```diff
@@ -0,0 +1,101 @@
+#include "wdog_imx.h"
+#include "board.h"
+
+#define NO_DOMAIN_PERM 0
```
Hmm, are these not also in soc.h?

galak commented on ext/hal/nxp/imx/utilities/inc/debug_console_imx.h:
```diff
@@ -0,0 +1,154 @@
+/*
```
Is there any point importing utilities/ since I'm not sure if we'd ever use debug_console.

galak commented on boards/arm/colibri_imx7d_m4/board.c:
```diff
@@ -0,0 +1,17 @@
+	return 0;
+}
+
+SYS_INIT(colibri_imx7d_init, PRE_KERNEL_1, 0);
```
Since this doesnt nothing, just remove board.c

galak commented on boards/arm/colibri_imx7d_m4/board.h:
```diff
@@ -0,0 +1,100 @@
+/* The board name */
+#define BOARD_NAME                            "IMX7_COLIBRI_M4"
+#define BOARD_DOMAIN_ID                       (1)
+
```
Why do we need all these defines?

galak commented on boards/arm/colibri_imx7d_m4/board.h:
```diff
@@ -0,0 +1,100 @@
+
+/* The board name */
+#define BOARD_NAME                            "IMX7_COLIBRI_M4"
+#define BOARD_DOMAIN_ID                       (1)
```
I think DOMAIN_ID should be a CONFIG param in the SoC and default to 1.  Than board code can change the default in Kconfig.defconfig if desired.

galak reviewed (dismissed):
Various code cleanups and questions.

diegosueiro commented:
@galak,

Any further feedback?

stanislav-poboril commented on ext/hal/nxp/imx/README:
```diff
@@ -0,0 +1,41 @@
+iMX7D Port
```
The drivers in this HAL will be used also for other SoC's (I am going to add iMX6SoloX's M4 core / Udoo Neo board in the near future). Therefore the iMX7D mentions should be changed to something more generic. Or I can change it once I add the other SoC support.

FYI, the package FreeRTOS_BSP_1.0.1_iMX6SX contains the following files in the platform/drivers folder, which are not in FreeRTOS_BSP_1.0.1_iMX7D:
- adc_imx6sx.h
- adc_imx6sx.c
- ccm_analog_imx6sx.h
- ccm_analog_imx6sx.c
- ccm_imx6sx.h
- ccm_imx6sx.c
- rdc_defs_imx6sx.h
- epit.h
- epit.c

FreeRTOS_BSP_1.0.1_iMX7D has the following specific files:
- adc_imx7d.h
- adc_imx7d.c
- ccm_analog_imx7d.h
- ccm_analog_imx7d.c
- ccm_imx7d.h
- ccm_imx7d.c
- rdc_defs_imx7d.h
- gpt.h
- gpt.c

Other differences in the platform/drivers folder are just some typos in the code comments.

I will add the missing files from FreeRTOS_BSP_1.0.1_iMX6SX as part of my future pull request, I think it should be placed in the same folder. But it will have multiple source URLs etc.

diegosueiro replied:
@stanislav-poboril,

> The drivers in this HAL will be used also for other SoC's (I am going to add iMX6SoloX's M4 core / Udoo Neo board in the near future). Therefore the iMX7D mentions should be changed to something more generic. Or I can change it once I add the other SoC support.

This is a great news. But as soon as I don't have an iMX6SoloX board I can't validate in this hardware and rename the files and configuration names to be more generic. I think that the best way of doing this is when you submit your changes you can make the appropriated generalizations.

stanislav-poboril replied:
Ok, I will do it along with my changes.

galak replied:
> The drivers in this HAL will be used also for other SoC's (I am going to add iMX6SoloX's M4 core / Udoo Neo board in the near future). Therefore the iMX7D mentions should be changed to something more generic. Or I can change it once I add the other SoC support.

@stanislav-poboril do you have something running on the Udoo Neo?  I was going to look at doing this next week, but if you've already got something I'll hold off.

stanislav-poboril replied:
@galak, I have hello_world running. I would like to rebase it against this pull request since it shares part of HAL. Then I would like to create a shim for Messaging Unit driver.

stanislav-poboril commented on ext/hal/nxp/imx/drivers/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+zephyr_sources(clock_freq.c)
+zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_IMX_CCM   ccm_imx7d.c)
+zephyr_sources_ifdef(CONFIG_CLOCK_CONTROL_IMX_CCM   ccm_analog_imx7d.c)
```
Should be also constrained by CONFIG_SOC_MCIMX7_M4, as CONFIG_SOC_MCIMX6X_M4 would add ccm_imx6sx.c and ccm_analog_imx6sx.c files instead.

diegosueiro replied:
Agreed.

stanislav-poboril commented on ext/hal/nxp/imx/drivers/clock_freq.c:
```diff
@@ -0,0 +1,243 @@
+/*
```
The files `clock_freq.h/c` are coming from examples folder in `FreeRTOS_BSP_1.0.1_iMX7D` and the content of those files is different in `FreeRTOS_BSP_1.0.1_iMX6SX`. It should be renamed or moved, maybe into `ext/hal/nxp/imx/devices/MCIMX7D/` folder?

diegosueiro replied:
Do you think that the functions in this module can be the same for i.MX7 and i.MX6SX?
If so, it doesn't make sense to move those files to `ext/hal/nxp/imx/devices/MCIMX7D/`

stanislav-poboril replied:
I haven't looked at the two set of files in much depth yet, but theirs diff is quite large.
The purpose of the files (getting frequency of SoC's particular subsystems) is similar to what can be found for example in `ext\hal\nxp\mcux\devices\MIMXRT1052\fsl_clock.c`, which is device specific.
So I am afraid there will have to be versions of clock_freq.h/c for each SoC based on FreeRTOS BSP (KSDK 1.x).

diegosueiro replied:
Ok. I'll move them to `ext/hal/nxp/imx/devices/MCIMX7D/`.

stanislav-poboril replied:
It also looks like the proper Zephyr way is to create abstraction of the device specific clock code (in `drivers/clock_control`) and use it instead of accessing the external API directly in driver shims. We may consider adding clock_control module for iMX6SX/7D in future.

diegosueiro replied:
Agreed. My idea is to have this basic support upstreamed and than implement other drivers (GPIO, I2C etc) and frameworks (clock_control, pinmux etc).

stanislav-poboril commented on drivers/serial/Kconfig.imx:
```diff
@@ -0,0 +1,73 @@
+# Kconfig - iMX7 M4 UART
```
iMX6SX uses the same driver in FreeRTOS_BSP_1.0.1_iMX6SX. Not sure whether to change the iMX7 mentions to something generic. Or I can mention the other SoC once I add its support.

diegosueiro replied:
In my opinion, all the generalization should be put in place when a new platform is going to be added.

stanislav-poboril replied:
Ok.

stanislav-poboril commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,146 @@
+ */
+
+#include <arm/armv7-m.dtsi>
+#include <dt-bindings/clock/imx_ccm.h>
```
This looks unused. The header is related to `drivers/clock_control/clock_control_mcux_ccm.c`, which has some minimum implementation to support iMX RT1050 (which uses `ext/hal/nxp/mcux`). The UART shim driver uses function from `clock_freq.c`.

diegosueiro replied:
I'll fix this.

stanislav-poboril commented on dts/arm/nxp/nxp_imx.dtsi:
```diff
@@ -0,0 +1,146 @@
+/*
```
I think the file `dts/arm/nxp/nxp_imx.dtsi` should be renamed to something more specific, for example `dts/arm/nxp/nxp_imx7d_m4.dtsi`. Unless we plan to list all cores in the file, in which case the `_m4` suffix would not be there (but here Zephyr will probably run only on M4 core, unlike in the case of `dts/arm/nxp/nxp_lpc54xxx.dtsi`, where Zephyr runs on both cores and the file lists them all).

(Now that we have multiple iMX SoC's, maybe also `dts/arm/nxp/nxp_rt.dtsi` could be renamed into something more specific, but it is not part of this pull request).

galak replied:
agreed, this should be imx7d-m4.dtsi.

diegosueiro committed:
ext/hal/nxp/imx: Import the nxp imx7 freertos bsp

This code component is used to add Zephyr support on iMX7 processors,
exclusively on Cortex M4 core, and to speed up the development process
it was decided to have it based on NXP FreeRTOS BSP implementation.

The source code was imported from the following folders:
    FreeRTOS_BSP_1.0.1_iMX7D/platform/drivers
    FreeRTOS_BSP_1.0.1_iMX7D/platform/devices

This source code depends on headers and sources from zephyr:
    ext/hal/cmsis

Origin: iMX7D NXP FreeRTOS BSP Peripheral Driver
License: BSD 3-Clause
URL: https://www.nxp.com/webapp/Download?colCode=FreeRTOS_iMX7D_1.0.1_LINUX&appType=license
commit: no commit hash
Purpose: The peripheral driver wraps the H/W for i.MX7 M4 core
Maintained-by: External

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
serial: Add imx uart driver shim

Adds a shim layer around the imx uart driver to adapt it to the Zephyr
serial interface.

Modem mode was introduce to control it as DCE and DTE and can be
configured in the device tree:
    modem-mode:
        type: int
        category: required
        description: Set the UART Port to modem mode 0 (dce) 1 (dte)
        generation: define

For now only the UART 2 was tested.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

stanislav-poboril commented on boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.dts:
```diff
@@ -0,0 +1,33 @@
+
+/dts-v1/;
+
+#include <nxp/nxp_imx.dtsi>
```
Please adjust the include after rename.

diegosueiro replied:
I have changed it in my local repo and for some reason the "boards/arm: add support for colibri_imx7d_m4 board " commit was not pushed to my zephry repo fork.

diegosueiro replied:
I think I forgot the last "git rebase --continue".

stanislav-poboril commented on boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.dts:
```diff
@@ -0,0 +1,33 @@
+#include <nxp/nxp_imx.dtsi>
+
+/ {
+	model = "TORADEX Colibir IMX7D board";
```
Typo - Colibir.

diegosueiro replied:
Thanks

stanislav-poboril commented on arch/arm/soc/nxp_imx/mcimx7_m4/Kconfig.defconfig.mcimx7_m4:
```diff
@@ -0,0 +1,40 @@
+endif # SERIAL
+
+config DOMAIN_ID
+    int
```
Please replace spaces by tab on this and the following line (checkpatch script fails on this).

diegosueiro committed:
arch: Add imx7d m4 soc support

The i.MX7 SoC is a Hybrid multi-core processor composed by Single/Dual
Cortex A7 core and Single Cortex M4 core.

Zephyr was ported to run on the M4 core. In a later release, it will
also communicate with the A7 core (running Linux) via RPmsg.

The low level drivers come from NXP FreeRTOS BSP and are located at
ext/hal/nxp/imx. More details can be found at ext/hal/nxp/imx/README

The A7 core is responsible to load the M4 binary application into the
RAM, put the M4 in reset, set the M4 Program Counter and Stack Pointer,
and get the M4 out of reset.
The A7 can perform these steps at bootloader level after the Linux
system has booted.

The M4 can use up to 5 different RAMs. These are the memory mapping for
A7 and M4:

+---------------+-----------------+---------------------------+
| Memory Name   | Start Address   | Size                      |
+===============+=================+===========================+
| TCML          | 0x007F8000      | 32KB                      |
+---------------+-----------------+---------------------------+
| TCMU          | 0x20000000      | 32KB                      |
+---------------+-----------------+---------------------------+
| OCRAM_S       | 0x20180000      | 32KB                      |
+---------------+-----------------+---------------------------+
| OCRAM         | 0x00900000      | 128KB                     |
+---------------+-----------------+---------------------------+
| DDR           | 0x10000000      | 256MB                     |
+---------------+-----------------+---------------------------+

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@MaureenHelm and @galak,

What are the next steps? Is there something else that I need to do?

MaureenHelm commented:
Hi @diegosueiro , can you please fix the CI failures? I looked at a few and they're overflowing the flash and ram sections. You can fix this by adding sizes to `boards/arm/colibri_imx7d_m4/colibri_imx7d_m4.yaml`, which will filter out CI tests that don't fit. See `boards/arm/frdm_kl25z/frdm_kl25z.yaml` for an example.

diegosueiro committed:
boards/arm: add support for colibri_imx7d_m4 board

The Colibri iMX7D Computer on Module with Colibri Evaluation Board
configuration supports the following hardware features on the Cortex M4
Core:

+-----------+------------+-------------------------------------+
| Interface | Controller | Driver/Component                    |
+===========+============+=====================================+
| NVIC      | on-chip    | nested vector interrupt controller  |
+-----------+------------+-------------------------------------+
| SYSTICK   | on-chip    | systick                             |
+-----------+------------+-------------------------------------+
| UART      | on-chip    | serial port-polling;                |
|           |            | serial port-interrupt               |
+-----------+------------+-------------------------------------+

The default configuration can be found in the defconfig file:
boards/arm/colibri_imx7d_m4/colibri_imx7d_m4_defconfig

Other hardware features are not currently supported by the port.

Connections and IOs:

The Colibri iMX7D Computer on Module with Colibri Evaluation Board
was tested with the following pinmux controller configuration.

+---------------+-----------------+---------------------------+
| Board Name    | SoC Name        | Usage                     |
+===============+=================+===========================+
| UART_B RXD    | UART2_TXD       | UART Console              |
+---------------+-----------------+---------------------------+
| UART_B TXD    | UART2_RXD       | UART Console              |
+---------------+-----------------+---------------------------+

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@MaureenHelm,

I added the memories sizes but the CI is still falling. It seems that it is not filtering by the flash size restrictions.

MaureenHelm commented:
> I added the memories sizes but the CI is still falling. It seems that it is not filtering by the flash size restrictions.

@diegosueiro The failing tests need to have the min_flash attribute added.

diegosueiro commented:
@MaureenHelm,

> @diegosueiro The failing tests need to have the min_flash attribute added.

Where should I add this? According to the documentation, min_flash and min_ram go into the tests yaml files. Is that correct?

galak commented:
> Where should I add this? According to the documentation, min_flash and min_ram go into the tests yaml files. Is that correct?
> 
> Should I create another PR to add the min_flash for those tests?

Its added in testcase.yaml.  For example you can look at tests/crypto/ecc_dh/testcase.yaml  which already has min_ram entry.  For example of min_flash, take a look at tests/crypto/mbedtls/testcase.yaml.

Yes, submit the adjustments to the testcases in a separate PR.

diegosueiro commented:
@galak,

How do I re-run the sanitycheck for this PR as soon as the PR #7008 was merged?

MaureenHelm commented:
recheck
